frontend/talentsearch/src/js/components/experienceForm/ExperienceSkills.tsx

```diff
@@ -0,0 +1,102 @@
+import React, { useEffect } from "react";
+import { useIntl } from "react-intl";
+import { useForm } from "react-hook-form";
+
+import type { Skill } from "../../api/generated";
+
+import AddSkillsToExperience from "../skills/AddSkillsToExperience/AddSkillsToExperience";
+import SkillsInDetail from "../skills/SkillsInDetail/SkillsInDetail";
+
+type FormValues = {
+  skills: { [id: string]: { details: string } };
+};
+
+export interface ExperienceSkillsProps {
+  skills: Skill[];
+  initialSkills?: Skill[];
+}
+
+const ExperienceSkills: React.FC<ExperienceSkillsProps> = ({
+  skills,
+  initialSkills,
+}) => {
+  const intl = useIntl();
+  const { setValue, watch } = useForm();
+  const watchedSkills = watch("skills");
+  const [addedSkills, setAddedSkills] = React.useState<Skill[]>([]);
+
+  const updateAddedSkills = () => {
+    if (watchedSkills) {
+      const newAddedSkills = skills.filter((skill) =>
+        Object.keys(watchedSkills).includes(skill.id),
+      );
+      setAddedSkills(newAddedSkills);
+    }
+  };
+
+  /**
+   * Updates an array of skills currently added to experience
+   */
+  useEffect(updateAddedSkills, [watchedSkills, skills]);
+
+  const handleAddSkill = (id: string) => {
+    const newSkills = { ...watchedSkills, [id]: { details: "" } };
+    setValue("skills", newSkills);
+  };
+
+  const handleRemoveSkill = (id: string) => {
+    const newSkills = Object.keys(watchedSkills).reduce(
+      (object: FormValues["skills"], key) => {
+        if (key !== id) {
+          object[key] = watchedSkills[key];
+        }
+        return object;
+      },
+      {},
+    );
+    setValue("skills", newSkills);
+  };
+
+  useEffect(() => {
+    let newSkills = {};
+    if (initialSkills) {
+      initialSkills?.forEach((s) => {
+        newSkills = {
+          ...newSkills,
+          [s.id]: {
+            details: s.experienceSkillRecord?.details || "",
+          },
+        };
+      });
+      setValue("skills", newSkills);
+    }
+  }, []);
```

Comment:
So, this is kind of a weird and it makes sense to some degree. I will be pushing up a commit to ignore this rule for this line but will explain why.

`useEffect` takes two arguments, a function and an array of dependencies. The function will fire every time one of the dependencies change between renders. The any use of a dependency within the function have the potential to change so it is recommended to be added to the array hence the rule. However, when you pass an empty dependency array, this function only ever fires on the first render (mounting) of the component.

For this particular case, this is simply massaging initial data to be used in the form values and setting them and since we only want this to ever run once, and we do not care if the dependency change since we only want it to run once and can safely ignore the rule.

I initially tried to just use default values for `react-hook-form`, but this doesn't seem to trigger an update where we are using `watch` to update components when the skills update.

Reply:
@vd1992 and I found when working on another PR that you need to pass initial data into the `watch` function separately. But you can do any data massaging once and pass the same data into `watch` and `defaultValues`.

Reply:
> @vd1992 and I found when working on another PR that you need to pass initial data into the `watch` function separately. But you can do any data massaging once and pass the same data into `watch` and `defaultValues`.

Wow! Thanks for the tip, still new with `react-hook-form` so I was unaware of this. I have updated it with [simplify watch default values](https://github.com/GCTC-NTGC/gc-digital-talent/pull/2480/commits/113d91412727477e69b74b03dad42c68f26c868b) and the whole thing is much cleaner now.